packages/gatsby/src/utils/jobs-manager.js

```diff
@@ -0,0 +1,203 @@
+const uuid = require(`uuid/v4`)
+const path = require(`path`)
+const hasha = require(`hasha`)
+const fs = require(`fs-extra`)
+const pDefer = require(`p-defer`)
+const slash = require(`slash`)
+const { createContentDigest } = require(`gatsby-core-utils`)
+const reporter = require(`gatsby-cli/lib/reporter`)
+const { store } = require(`../redux`)
+
+let activityForJobs
+let activeJobs = 0
+
+/** @type {Map<string, {id: string, deferred: pDefer.DeferredPromise}>} */
+const jobsInProcess = new Map()
+
+/**
+ * @param {string} path
+ * @param {string} rootDir
+ * @return {string}
+ */
+const convertPathsToRelative = (filePath, rootDir) => {
+  const relative = path.relative(rootDir, filePath)
+
+  if (relative.includes(`..`)) {
```

Comment:
I'm just not allowing any relative paths that aren't in the program root. The path will always be  correct when using `path.relative` as it resolves to the full path so ~ becomes /home/<user> and will have ../ inside. Locally this won't break ever but it will break in cloud. That's why this check exists.